### PR TITLE
Implemented Merge Item Feature

### PR DESCRIPTION
**Pull request by cydh**
- [ ] Code review
- [ ] QA testing on idRO
- [ ] QA testing on iRO
- Merge item only will be available only after talk with certain NPC (such Mergician)
- Received packets:
  - `'096D' => ['merge_item_open', 'v a*', [qw(length itemList)]], #-1`
  - `'096F' => ['merge_item_result', 'v v C', [qw(itemIndex total result)]], #5`
- Sent packets:
  - `'096E' => ['merge_item_request', 'v a*', [qw(length itemList)]], #-1`
  - `'0974' => ['merge_item_cancel'], #2`
- Added global array var: `@mergeItemList`
- Added command:
  - `merge`: Only can be used atfer server allows it.
- Thanks to @rAthena for packet data

Signed-off-by: Cydh Ramdh cydh@pservero.com

### CODE REVIEW

**Comment by cydh** [comment]
I'll rebase this

File: src/Network/Send.pm [hunk 1]
```diff
@@ -1221,4 +1221,42 @@ sub sendProduceMix {
...
+# Request to merge item
+# 096E <size>.W { <index>.W }*
+# @author [Cydh]
+##
+sub sendMergeItemRequest {
+	my ($self, $num, $items) = @_;
+	my $len = ($num * 4) + 12;
+	$self->sendToServer($self->reconstruct({
+		switch => 'merge_item_request',
+		len => $len,
```
Comment (farrainbow) [inline src/Network/Send.pm]: Packet length should automatically be added (if packet lengths are configured properly) by `reconstruct`, no need to specify it:

https://github.com/OpenKore/openkore/blob/a668f7f8701974a995b51dfa9985631113bb8096/src/Network/PacketParser.pm#L195

File: src/Network/Receive.pm [hunk 2]
```diff
@@ -2406,4 +2406,48 @@ sub hat_effect {
...
+##
+sub merge_item_open {
+	my ($self, $args) = @_;
+	my $num = ($args->{length} - 4) / 2;
+	message TF("Received %d items to be merged. Use 'merge' to continue\n", $num), "info";
+}
+
+sub parse_merge_item_open {
+	my ($self, $args) = @_;
+	@mergeItemList = map {{ itemIndex => unpack('v', $_) }} unpack '(a2)*', $args->{itemList};
```
Comment (farrainbow) [inline src/Network/Receive.pm]: The original idea was for `parse_*` functions to just further unpack `$args` and do nothing except that (do not modify the global state).

For example, `$args->{items} = [map ... $args->{itemList}]` in `parse_merge_item_open` and then assign `@mergeItemList = @{$args->{items}}` in actual `merge_item_open` handler.

File: src/Network/Receive.pm [hunk 3]
```diff
@@ -2406,4 +2406,48 @@ sub hat_effect {
 	}
 }
 
+##
+# 096D <size>.W { <index>.W }*
+# @author [Cydh]
+##
+sub merge_item_open {
+	my ($self, $args) = @_;
+	my $num = ($args->{length} - 4) / 2;
```
Comment (farrainbow) [inline src/Network/Receive.pm]: Maybe just use `$num = scalar @mergeItemList` here, displays intent better, and `length` would not be needed at all anywhere.

File: src/Network/Send.pm [hunk 4]
```diff
@@ -1221,4 +1221,42 @@ sub sendProduceMix {
...
+sub sendMergeItemCancel {
+	my ($self, $args) = @_;
+	$self->sendToServer($self->reconstruct({ switch => 'merge_item_cancel' }));
+	debug "Cancel Merge item\n", "sendPacket";
+	undef @mergeItemList;
+}
+
+sub reconstruct_merge_item_cancel {
+	my ($self, $args) = @_;
+}
```
Comment (farrainbow) [inline src/Network/Send.pm]: If any `parse`/`reconstruct` functions are empty, there's no need to include them at all. They're called only when they do exist:

https://github.com/OpenKore/openkore/blob/a668f7f8701974a995b51dfa9985631113bb8096/src/Network/PacketParser.pm#L185

File: src/Commands.pm [hunk 5]
```diff
@@ -5997,4 +6000,91 @@ sub cmdDeadTime {
...
+				$it->{item} = $item;
+				$it->{invIndex} = $item->{invIndex};
+				$msg .= swrite(
+					"@<<< @<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<",
+					[$item->{invIndex}, $display]);
+			}
+		}
+		$msg .= ('-'x50) . "\n";
+		message $msg, "list";
+		message T("Usages: merge <item #>,<item #>[,<item #>,<item #>,...]\n"), "info";
```
Comment (farrainbow) [inline src/Commands.pm]: "Usage" would be proper, maybe?

**Comment by kaliwanagan** [comment]
@cydh are you alright with the suggestions above?

**Comment by moneyisgod** [comment]
bump

**Comment by cydh** [comment]
rebased and `Allows edit from maintainers.`
I can't apply the reviews yet, not sure when I can test so I leave it as be. Maybe later by using rAthena, but idk when. feel free if anyone want to merge/fetch this to other branch before merge this to master.

**Comment by Nipodemos** [comment]
@cydh anything else you would like to add before merging?

**Comment by cydh** [comment]
@Nipodemos I forgot to retest in idRO after this branch update, I'll test ti tonight

**Comment by Nipodemos** [comment]
Okay, just waiting for your reply

**Comment by cydh** [comment]
just tested in idRO, working good
merge-item-5

**Pull request merged**